internal/provider/waypoint/data_source_waypoint_application_template.go

```diff
@@ -0,0 +1,227 @@
+	waypoint_models "github.com/hashicorp/hcp-sdk-go/clients/cloud-waypoint-service/preview/2023-08-18/models"
+	"github.com/hashicorp/terraform-plugin-framework-validators/datasourcevalidator"
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-provider-hcp/internal/clients"
+)
+
+var _ datasource.DataSource = &DataSourceApplicationTemplate{}
+var _ datasource.DataSourceWithConfigValidators = &DataSourceApplicationTemplate{}
+
+func (d DataSourceApplicationTemplate) ConfigValidators(ctx context.Context) []datasource.ConfigValidator {
+	return []datasource.ConfigValidator{
+		datasourcevalidator.Conflicting(
+			path.MatchRoot("name"),
+			path.MatchRoot("id"),
+		),
+	}
+}
+
+type DataSourceApplicationTemplate struct {
+	client *clients.Client
+}
+
+type DataSourceApplicationTemplateModel struct {
+	ID                     types.String `tfsdk:"id"`
+	Name                   types.String `tfsdk:"name"`
+	ProjectID              types.String `tfsdk:"project_id"`
+	OrgID                  types.String `tfsdk:"organization_id"`
+	Summary                types.String `tfsdk:"summary"`
+	Labels                 types.List   `tfsdk:"labels"`
+	Description            types.String `tfsdk:"description"`
+	ReadmeMarkdownTemplate types.String `tfsdk:"readme_markdown_template"`
+
+	TerraformCloudWorkspace *tfcWorkspace    `tfsdk:"terraform_cloud_workspace_details"`
+	TerraformNoCodeModule   *tfcNoCodeModule `tfsdk:"terraform_no_code_module"`
+}
+
+func NewApplicationTemplateDataSource() datasource.DataSource {
+	return &DataSourceApplicationTemplate{}
+}
+
+func (d *DataSourceApplicationTemplate) Metadata(ctx context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_waypoint_application_template"
+}
+
+func (d *DataSourceApplicationTemplate) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "The Waypoint Template data source retrieves information on a given Application Template.",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed:    true,
+				Optional:    true,
+				Description: "The ID of the Application Template.",
+			},
+			"name": schema.StringAttribute{
+				Description: "The name of the Application Template.",
+				Computed:    true,
+				Optional:    true,
+			},
+			"organization_id": schema.StringAttribute{
+				Description: "The ID of the HCP organization where the Waypoint Application Template is located.",
+				Computed:    true,
+			},
+			"project_id": schema.StringAttribute{
+				Description: "The ID of the HCP project where the Waypoint Application Template is located.",
+				Optional:    true,
```

Comment:
I feel like this might be required? I'm not sure though, but what happens when you try to use an application template resource without an HCP project ID?

Reply:
I think the validator will reject the application template and return an error if there is no project ID

Reply:
What @HenryEstberg said; optional here means the user can optionally provide the `project_id` if they want to create this template for a different project than the one their provider is initially configured to work with. A default is obtained from the client ID/Secret pair. In the future, RBAC may prevent this from working in some cases, but for now it's recommended by the HCP provider team(?) to allow for an optional `project_id`